### PR TITLE
feature: Add optional port to host mappings.

### PR DESCRIPTION
**Pull request by calavera**
This is an extension for the host mapping feature. It allows people to map to a different port than the one in the original url.

This is very convenient when you run k6 against a fleet of docker
containers to test the same service but want to keep the port mapping.

Signed-off-by: David Calavera <david.calavera@gmail.com>

### CODE REVIEW

File: lib/testutils/httpmultibin/httpmultibin.go
Comment (MStoykov) [inline lib/testutils/httpmultibin/httpmultibin.go:253]: Possibly outside of this PR as I don't think it's worth it, but with the new "hosts" we can drop the `_PORT` and `_URL` varaints

File: lib/netext/dialer_test.go
Comment (MStoykov) [inline lib/netext/dialer_test.go:32]: can you combine the two tests, by adding another domain without a "binding" without a port so it gets resolved by the testResolver instead of the `hosts` binding? 

Also maybe have a test of the `blacklist`  in combination with `hosts` ? Maybe just a third domain with default resolving to a blacklisted IP, a `hosts` binding that is blacklisted again and one that isn't?

File: lib/netext/dialer.go
Comment (MStoykov) [inline lib/netext/dialer.go:93]: I guess this was something to do with IPv6 ? maybe have a test with it IPv6 as well?

**Comment by calavera** [comment]
> So unfortunately this (as well as a bunch of other PRs) will probably wait for at least another month (hopefully not more) until we get back to you

Thanks for the heads up @MStoykov. Don't worry about it, take your time with all other priorities that you have. I actually wrote this code months ago an never had time to push the PR. I'll address your comments for when you're ready to take a deeper look.

I was going to write some documentation about this change, but I couldn't find any place to do it. Any pointers will be appreciated.

**Comment by MStoykov** [comment]
> I was going to write some documentation about this change, but I couldn't find any place to do it. Any pointers will be appreciated.

I meant to spare you that ... but, given that you asked :D , there is this [k6-docs](https://github.com/loadimpact/k6-docs) repo :D.

 Also if you click on the `suggest edits` on top of [this page](https://k6.io/docs/using-k6/options#hosts) you will be forwarded to the actual file you need to change (as the structure isn't very intuitive at least for me :( )

Thanks again for all the work :D

**Comment by calavera** [comment]
@MStoykov I've updated the code with your suggestions, thanks! :raised_hands:

I'll wait for your review to make any changes in the docs.

File: lib/netext/dialer.go
Comment (calavera) [reply lib/netext/dialer.go:93]: I discovered that these changes don't work very well with IPv6 hosts. I'll change the logic to make it work and add new tests.

Comment (calavera) [reply lib/netext/dialer.go:93]: I've added new tests for ipV6.

**Comment by MStoykov** [comment]
Hi @calavera , we have released v0.27.0 last week, and while we are fixing up some stuff for v0.27.1, you should probably rebase this after we merge #1562 as that will have confilcts. 
Hopefully we will be able to get this in v0.28.0 ;)

**Comment by calavera** [comment]
Thanks for the update @MStoykov :+1:

File: lib/netext/dialer_test.go
Comment (calavera) [reply lib/netext/dialer_test.go]: @imiric thanks for the link. I've incorporated some of those cases, which helped me spot a bug, and I've updated the PR.

File: lib/netext/dialer.go
Comment (imiric) [inline lib/netext/dialer.go:189]: You could try `ParseIP()` right after calling `SplitHostPort()`, and if it is an IP you can skip the DNS lookup, which would avoid returning this error masking as a resolver error. See what I mean in 4c0e6639.

Comment (calavera) [reply lib/netext/dialer.go:192]: that's a great idea. I've pushed a new commit with that change.

File: lib/options.go
Comment (MStoykov) [inline lib/options.go:257]: are you here trying to figure out if there is a port ? so that `net.SplitHostPort` doesn't return you an error?  Because IMO you can just call it - and if there is no error - there is a port, if there is an error - there is no port :D 
As far as I can see this code currently doesn't handle `something.com:8413` for example correctly ?

File: lib/netext/dialer.go
Comment (calavera) [reply lib/netext/dialer.go]: I've moved some code to make this more clear. The first call checks if the full address is in the Hosts map, the second call checks if only the host part is in the Hosts map. Let me know if that doesn't make sense.

File: lib/options.go
Comment (calavera) [reply lib/options.go]: I've removed the double check.

>  As far as I can see this code currently doesn't handle something.com:8413 for example correctly ?

I don't understand this commend. As far as I can tell, the current code expects the option to be a valid IP address, and `something.com` won't ever work as IP address.

File: lib/netext/dialer.go
Comment (MStoykov) [inline lib/netext/dialer.go:144]: This doesn't dial and address, maybe `getDialAddr` or `toDialAddr`

File: lib/netext/dialer.go
Comment (MStoykov) [inline lib/netext/dialer.go:151]: I find this cast ... strange ... maybe we should change `lib.IPNet` so it isn't required
cc @na-- @imiric

File: lib/netext/dialer.go
Comment (MStoykov) [inline lib/netext/dialer.go:193]: won't it be easier to  be just 

```
if remote == nil {
return nil,nil
}
```

Also I am not really certain the name is accurate ... maybe `checkInHosts`? and maybe it should return a bool which is `true` if the check was possitive

Comment (MStoykov) [reply lib/netext/dialer.go:193]: Also looking at this should this whole function be 
```
if remote, ok :=d.Hosts[addr]; ok{
  return remote, nil
}
if remote, ok := d.Hosts[host]; ok {
// here is the part which adds the port , which also could possibly be a integer here ? so you don't need to have a change for an error ?
}
return nil,nil
```
Because currently even if the whole `addr` matches you will make a new HostAddress ?

**Comment by calavera** [comment]
Thanks for the review @MStoykov. I pushed new changes.

File: lib/netext/dialer.go
Comment (MStoykov) [inline lib/netext/dialer.go:202]: As the test has shown here you get a pointer to the remote which you then ... change, which definitely should never happen given that this is the configuration ;)

File: lib/netext/dialer.go
Comment (MStoykov) [inline lib/netext/dialer.go:196]: `getCachedHost` should probably be `getOverwrittenHost` or `getConfiguredHost`, as there is nothing cached it is literally configured/overwritten in the settings ;)

File: lib/netext/dialer.go
Comment (MStoykov) [inline lib/netext/dialer.go:198]: this and the above comment are not necessary IMO and are also wrong from the context of just this function body - you don't resolve it through DNS in here

File: lib/netext/dialer.go
Comment (calavera) [reply lib/netext/dialer.go:196]: changed

File: lib/netext/dialer.go
Comment (imiric) [inline lib/netext/dialer.go:179]:
```suggestion
func (d *Dialer) fetchRemoteFromResolver(host, port string) (*lib.HostAddress, error) {
```

File: lib/netext/dialer_test.go
Comment (imiric) [inline lib/netext/dialer_test.go:4]:
```suggestion
 * Copyright (C) 2020 Load Impact
```

**Review by imiric — APPROVED** [review]
This LGTM, save for the typo nitpick. Thanks for your work @calavera and apologies about the delay with merging this.

**Comment by calavera** [comment]
I've fixed those typos, thanks for the review!

**Review by imiric — APPROVED** [review]

**Review by na-- — APPROVED** [review]
LGTM, though we should manually test this with `k6 archive` (for cloud issues) and envconfig issues (`K6_HOSTS`)

**Review by MStoykov — APPROVED** [review]

**Pull request merged**
